File: c++/basics/2f_design_browser_history.cpp
// Design Browser history with methods to go to prev and next pages/urls.

class Node{
    public: 
        string val;
        Node* prev;
        Node* next;
        
        Node(string val){
            this->val = val;
            this->prev = NULL;
            this->next = NULL;
        }
};

class BrowserHistory {
public:
    Node* curr;
    BrowserHistory(string homepage) {
        curr = new Node(homepage); 
    }
    
    void visit(string url) {
        Node* fwd = new Node(url); 
        curr->next = fwd;
        fwd->prev = curr;
        curr = curr->next; 
    }
    
    string back(int steps) {
        while(curr->prev && steps>0){
            curr = curr->prev;
            steps--;
        }
        return curr->val;
    }
    
    string forward(int steps) {
        while(curr->next && steps>0){
            curr = curr->next;
            steps--;
        }
        return curr->val;
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */

File: js/basics/2f_design_browser_history.js
// Design Browser history with methods to go to prev and next pages/urls.

/**
 * @param {string} homepage
 */

var Node = function(val, prev, next) {
    this.val = (val === undefined ? 0 : val);
    this.prev = (prev === undefined ? null : prev);
    this.next = (next === undefined ? null : next);
}

var BrowserHistory = function(homepage) {
    this.curr = new Node(homepage)
};

/** 
 * @param {string} url
 * @return {void}
 */
BrowserHistory.prototype.visit = function(url) {
    let newNode = new Node(url)
    this.curr.next = newNode 
    newNode.prev = this.curr
    this.curr = this.curr.next 
};

/** 
 * @param {number} steps
 * @return {string}
 */
BrowserHistory.prototype.back = function(steps) {
    while(this.curr.prev && steps>0){
        this.curr = this.curr.prev
        steps--
    }
    return this.curr.val
};

/** 
 * @param {number} steps
 * @return {string}
 */
BrowserHistory.prototype.forward = function(steps) {
    while(this.curr.next && steps>0){
        this.curr = this.curr.next
        steps--
    }
    return this.curr.val
};

/** 
 * Your BrowserHistory object will be instantiated and called as such:
 * var obj = new BrowserHistory(homepage)
 * obj.visit(url)
 * var param_2 = obj.back(steps)
 * var param_3 = obj.forward(steps)
 */